sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/timewindow/recurrence/RecurrenceEvaluator.java

```diff
@@ -0,0 +1,848 @@
+        ZonedDateTime date = firstDateOfWeek;
+        ZonedDateTime prevOccurrence = firstDateOfWeek;
+        Duration minGap = Duration.ofDays(7);
+        for (int i = 0; i < 6; i++) {
+            date = date.plusDays(1);
+            final ZonedDateTime finalDate = date;
+            if (daysOfWeek.stream().anyMatch(dayOfWeek -> convertToWeekDayNumber(dayOfWeek) == getDayOfWeek(finalDate))) {
+                final Duration currentGap = Duration.between(prevOccurrence, date);
+                if (currentGap.compareTo(minGap) < 0) {
+                    minGap = currentGap;
+                }
+                prevOccurrence = date;
+            }
+        }
+        if (interval == 1) {
+            // It may across weeks. Check the adjacent week
+            date = date.plusDays(1);
+            final Duration currentGap = Duration.between(prevOccurrence, date);
+            if (currentGap.compareTo(minGap) < 0) {
+                minGap = currentGap;
+            }
+        }
+        return minGap.compareTo(duration) >= 0;
+    }
+
+    private int remainingDaysOfWeek(int today, int firstDayOfWeek) {
+        int remainingDays = (today - firstDayOfWeek);
+        if (remainingDays < 0) {
+            return -remainingDays;
+        } else {
+            return RecurrenceConstants.WEEK_DAY_NUMBER - remainingDays;
+        }
+    }
+
+    private boolean tryValidateAbsoluteMonthlyRecurrencePattern() {
+        final RecurrencePattern pattern = settings.getRecurrence().getPattern();
+
+        // Time window duration must be shorter than how frequently it occurs
+        final Duration intervalDuration = Duration.ofDays((long) pattern.getInterval() * RecurrenceConstants.MIN_MONTH_DAY_NUMBER);
+        if (!tryValidateTimeWindowDuration(settings, intervalDuration)) {
+            return false;
+        }
+
+        // Required parameters
+        if (!tryValidateDayOfMonth(settings)) {
+            return false;
+        }
+
+        // Check whether "Start" is a valid first occurrence
+        final ZonedDateTime alignedStart = settings.getStart().withZoneSameInstant(getRecurrenceTimeZoneId());
+        if (alignedStart.getDayOfMonth() != pattern.getDayOfMonth()) {
+            paramName = TIME_WINDOW_FILTER_SETTING_START;
+            reason = RecurrenceConstants.NOT_MATCHED;
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean tryValidateDayOfMonth(TimeWindowFilterSettings settings) {
+        paramName = String.format("%s.%s.%s", TIME_WINDOW_FILTER_SETTING_RECURRENCE, RecurrenceConstants.RECURRENCE_PATTERN, RecurrenceConstants.RECURRENCE_PATTERN_DAY_OF_MONTH);
+        if (settings.getRecurrence().getPattern().getDayOfMonth() == null) {
+            reason = RecurrenceConstants.REQUIRED_PARAMETER;
+            return false;
+        }
+        if (settings.getRecurrence().getPattern().getDayOfMonth() < 1 || settings.getRecurrence().getPattern().getDayOfMonth() > 31) {
```

Comment:
Not that this isn't complicated enough, but are we just going to assume there are 31 days in every month, for validation?

Reply:
If the patter is "absolute yearly", then the user will set both month and dayOfMonth, we should set the max day based on the month number. For example, "month: 9, dayOfMonth: 31" is not allowed.
If the pattern is "absolute monthly", which means the enabled period may repeat once a month. We should use 31 as the max number.

Reply:
> If the patter is "absolute yearly", then the user will set both month and dayOfMonth, we should set the max day based on the month number. For example, "month: 9, dayOfMonth: 31" is not allowed. If the pattern is "absolute monthly", which means the enabled period may repeat once a month. We should use 31 as the max number.

What is the intended result when 31 is selected for absolute monthly in a month with only 30 days for less? Will it run on the 1st or the equivalent, or skip that month?